phabricatoremails/progress.py

```diff
@@ -0,0 +1,58 @@
+import sys
+import time
+from dataclasses import dataclass
+from logging import Logger
+from typing import Callable, Any
+
+
+@dataclass
+class PhabricatorProgress:
+    _logger: Logger
+    _poll_gap_seconds: int
+
+    def process(self, db, process_next: Callable[[Any, str], str]):
+        while True:
+            with db:
+                with db.cursor() as cursor:
+                    cursor.execute("SELECT up_to_key FROM query_position")
+                    query_position = cursor.fetchone()
+                    if not query_position:
+                        self._logger.error(
+                            'The "query_position" table is not populated, '
+                            'try running "phabricatoremails init"?'
+                        )
+                        sys.exit(1)
+
+                    from_key = query_position["up_to_key"]
+                    last_key = process_next(cursor, from_key)
+
+                    if from_key == last_key:
+                        self._logger.debug(
+                            "Caught up with feed at key {}, sleeping for {} seconds...".format(
+                                last_key, self._poll_gap_seconds
+                            )
+                        )
+                        time.sleep(self._poll_gap_seconds)
+                        continue
+
+                    cursor.execute(
+                        "UPDATE query_position SET up_to_key = %s", (last_key,)
+                    )
+
+    @staticmethod
+    def set_initial_position(cursor, status: int):
+        cursor.execute("INSERT INTO query_position (up_to_key) VALUES (%s)", (status,))
+
+
+class StaticProgress:
+    def __init__(self, key):
+        self._key = key
+
+    @staticmethod
+    def set_initial_position(cursor, status: int):
```

Comment:
What is the point of having this method here? If it is meant to be consistent with the above method, perhaps creating an abstract base class would help.

Reply:
I am so excited for us to have Python 3.8 since it adds ["protocols"/"structual subtyping"](https://www.python.org/dev/peps/pep-0544/), which is _exactly_ what we're seeing here with `set_initial_position(...)`.

But yeah, this method exists because it's being used via "duck typing" - it's used in [`init`](https://github.com/mozilla-conduit/phabricator-emails/blob/1609266-phab-emails/phabricatoremails/init.py#L28). When using `PhabricatorProgress`, then we should set the initial position in the database. When we're developing locally and we're using `StaticProgress`, then we want to ignore the position we're provided.